pose/src/posenet/teachable-posenet.ts

```diff
@@ -38,25 +46,52 @@ const isTensor = (c: any): c is tf.Tensor =>
  * the data as a JS Array.
  */
 function flatOneHot(label: number, numClasses: number) {
-    const labelOneHot = new Array(numClasses).fill(0);
+    const labelOneHot = new Array(numClasses).fill(0) as number[];
     labelOneHot[label] = 1;
+
     return labelOneHot;
 }
 
-function convertToTfDataset(xs: Float32Array[], ys: number[][]) {
-    const xTrain = tf.data.array(xs);
-    const yTrain = tf.data.array(ys);
-    const trainDataset = tf.data.zip({ xs: xTrain,  ys: yTrain});
-    //TODO: ys.length might not always be best
-    const shuffled = trainDataset.shuffle(ys.length);
+/**
+ * Shuffle an array of Float32Array or Samples using Fisher-Yates algorithm
+ * Takes an optional seed value to make shuffling predictable
+ */
+function fisherYates(array: Float32Array[] | Sample[], seed?: seedrandom.prng) {
+    const length = array.length;
+
+    // need to clone array or we'd be editing original as we goo
+    let shuffled = array.slice();
+
+    for (let i = (length - 1); i > 0; i -= 1) {
+        let randomIndex ;
+        if (seed) {
+            randomIndex = Math.floor(seed() * (i + 1));
+        }
+        else {
+            randomIndex = Math.floor(Math.random() * (i + 1));
+        }
+        
+        [shuffled[i], shuffled[randomIndex]] = [shuffled[randomIndex],shuffled[i]];
+    }
+  
     return shuffled;
 }
 
-export class TeachablePoseNet extends CustomPoseNet{
-    // Array<[className, activation]>
-    public examples: Array<[number, Float32Array]> = [];
-    private ys: number[][];
-    private dataset: tf.data.Dataset<TensorContainer>;
+export class TeachablePoseNet extends CustomPoseNet {
+    /**
+     * Training and validation datasets
+     */
+    private trainDataset: tf.data.Dataset<TensorContainer>;
+    private validationDataset: tf.data.Dataset<TensorContainer>;
+
+    // Number of total samples
+    private totalSamples: number = 0;
+
+    // Array of all the examples collected
+    public examples: Array<Array<Float32Array>> = [];
+
+    // Optional seed to make shuffling of data predictable
+    private seed: seedrandom.prng;
```

Comment:
👍 This is super helpful!

Is the end-to-end model creation process still non-deterministic?  I'm wondering if there are things within `#fit` like random initialization and sampling within the SGD process that don't share the same seed param as here.

Reply:
This seed is just for our shuffling of the dataset before and after splitting into training and validation (and only on the Teachable Machine library end). But we actually don't use it in Teachable Machine the app proper. More so for testing. 

Tensorflow.js the library happens to use seeds for a number of functions you can find them by just searching the [API](https://js.tensorflow.org/api/latest/) for seed.

Reply:
Yep, gotcha!  This is just for testing this library in isolation.

To confirm my understanding of TM, if I train a TM image classification model with the same data set, the data shuffling and splitting will be non-deterministic, and the initialization and fitting process within TF will also be non-deterministic, right?

In other words, if a class of 10 students train the same model on the same data set, we would expect that they would learn different models that would perform differently on a separate generalization data set.

Reply:
Yep, correct. Although the difference should not be large if all other settings remain equal OR if the data within each class is somewhat similar. 

I think currently we have the sed on in the live version of the app for testing.

Reply:
got it, thanks! 👍

Reply:
I should add though that [Math.random()](https://v8.dev/blog/math-random) is only a pseudo-random number generator, even if we don't use a seed.